Teach dvsim to understand BUILD_ROOT

rswarbrick opened the pull request:
This should close #5031. It also does a bit of other tidying up to make the code in question a bit easier to follow.

We're not actually using $BUILD_ROOT anywhere in our tree at the moment, but this needs to land first to make that do anything. And I checked it locally, so I'm reasonably convinced this works!

rswarbrick committed:
[dvsim] Don't try to work around NFS latency in dvsim.py

There was some code that tried to do `ls -d` in the scratch root, with
a 1 second timeout. The idea was that we'd silently work in a faster
location if the directory was a bit slow. It's neat, but probably a
source of confusion, and it's probably better for the user to set
their working directory explicitly.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick committed:
[dvsim] Remove some unneeded error handling in dvsim.py

This code tried to print out something helpful if the user had tried
to access "/foo/bar/baz" by accident. But the default message you get
anyway seems equally helpful. Let's remove the unnecessary complexity.

  $ util/dvsim/dvsim.py --scratch-root=/root/foo hw/ip/uart/dv/uart_sim_cfg.hjson
  Traceback (most recent call last):
    File "/home/rjs/work/opentitan/util/dvsim/dvsim.py", line 776, in <module>
      main()
    File "/home/rjs/work/opentitan/util/dvsim/dvsim.py", line 688, in main
      args.scratch_root = resolve_scratch_root(args.scratch_root, proj_root)
    File "/home/rjs/work/opentitan/util/dvsim/dvsim.py", line 66, in resolve_scratch_root
      os.makedirs(arg_scratch_root, exist_ok=True)
    File "/nix/store/wm85wvkj8w9lk3s53k6s3zkgi5p4xf2j-python3-3.10.17/lib/python3.10/os.py", line 225, in makedirs
      mkdir(name, mode)
  PermissionError: [Errno 13] Permission denied: '/root/foo'

Seems pretty clear to me!

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick committed:
[dvsim] Remove check that scratch root is writable in dvsim.py

Without this check, you now get something like this:

  $ util/dvsim/dvsim.py hw/top_earlgrey/dv/chip_sim_cfg.hjson --scratch-root=/root
  INFO: [dvsim] [proj_root]: /home/rjs/work/opentitan
  Traceback (most recent call last):
    File "/home/rjs/work/opentitan/util/dvsim/dvsim.py", line 751, in <module>
      main()
    File "/home/rjs/work/opentitan/util/dvsim/dvsim.py", line 672, in main
      (Path(args.scratch_root) / 'FUSESOC_IGNORE').touch()
    File "/usr/lib/python3.11/pathlib.py", line 1109, in touch
      fd = os.open(self, flags, mode)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  PermissionError: [Errno 13] Permission denied: '/root/FUSESOC_IGNORE'

which seems reasonable to me. Note that if we were really
enthusiastic, we could probably have fooled the existing
check (suppose the directory is writable, but the file we need to
create exists and is read-only). Probably no point in trying this at
all.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick committed:
[dvsim] Shuffle code in dvsim's resolve_scratch_root

If we move the mkdirs call to the callsite, we can use 'return' to
simplify the early exit behaviour. This also avoids the thing called
"resolve" changing the filesystem, which I think was a bit confusing.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick committed:
[dvsim] Teach dvsim to obey $BUILD_ROOT

This still uses the command line argument (--scratch-root) by
preference, or SCRATCH_ROOT if it's defined. But it now tries
BUILD_ROOT/build-scratch before falling back to the project root,
resolving issue #5031.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick commented:
This is a clone of #18030 (which I messed up reopening).

machshev reviewed (approved):
Looks good to me.

Nit pick on `os.path` verses `pathlib` but looks functionally sound. Also potentially changing behavior by no longer supporting relative paths in scratch dir path?

rswarbrick committed:
[dvsim] Make type dispatch in _stringify_wildcard_value cleaner

Also, support the Path type as a "special sort of string", which
allows slightly stronger typing elsewhere in the tool.

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

rswarbrick commented on util/dvsim/dvsim.py:
```diff
@@ -50,47 +49,28 @@
+        if not root:
+            continue
+        if subdir:
+            return os.path.join(root, subdir)
```
You're right, and the result is much nicer. Thanks!

rswarbrick committed:
[dvsim] Use string interpolation in a place in FlowCfg.py

This allows self.scratch_root to be a Path object instead of a
string (which will happen with the next commit).

Signed-off-by: Rupert Swarbrick <rswarbrick@lowrisc.org>

machshev reviewed (approved):
Nice, looks good to me!
